### PR TITLE
8341927: Replace hardcoded security providers with new test.provider.name system property

### PR DESCRIPTION
**Pull request by mpdonova**
In this PR, I removed hard-coded security providers and replaced them with a system property, test.provider.name. If the property is not specified, the provider originally used in the test is used:

Cipher c = Cipher.getInstance("AES/GCM/NoPadding", System.getProperty("test.provider.name", "SunJCE"));

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8341927](https://bugs.openjdk.org/browse/JDK-8341927): Replace hardcoded security providers with new test.provider.name system property (**Enhancement** - P2)

### Reviewers
 * [Anthony Scarpino](https://openjdk.org/census#ascarpino) (@ascarpino - **Reviewer**) Review applies to [298670c6](https://git.openjdk.org/jdk/pull/21551/files/298670c6731d7e6e05cf0965c34a714bacc07431)
 * [Sean Mullan](https://openjdk.org/census#mullan) (@seanjmullan - **Reviewer**)
 * [Rajan Halade](https://openjdk.org/census#rhalade) (@rhalade - **Reviewer**) Review applies to [3889e339](https://git.openjdk.org/jdk/pull/21551/files/3889e33963607b213b719fc0153fea85650b3e0c)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21551/head:pull/21551` \
`$ git checkout pull/21551`

Update a local copy of the PR: \
`$ git checkout pull/21551` \
`$ git pull https://git.openjdk.org/jdk.git pull/21551/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21551`

View PR using the GUI difftool: \
`$ git pr show -t 21551`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21551.diff">https://git.openjdk.org/jdk/pull/21551.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21551#issuecomment-2417669485)

### CODE REVIEW

File: test/jdk/javax/security/auth/login/Configuration/GetInstance.java [hunk 1]
```diff
@@ -84,12 +84,14 @@ private int testDefault(int testnum) throws Exception {
 
     private int testStringProvider(int testnum) throws Exception {
         // get an instance of JavaLoginConfig from SUN
-        Configuration c = Configuration.getInstance(JAVA_CONFIG, null, "SUN");
+        Configuration c = Configuration.getInstance(JAVA_CONFIG, null,
+                System.getProperty("test.provider.name","SUN"));
```
Comment (turbanoff) [inline test/jdk/javax/security/auth/login/Configuration/GetInstance.java:88]:
```suggestion
                System.getProperty("test.provider.name", "SUN"));
```

File: test/jdk/javax/security/auth/login/Configuration/GetInstance.java [hunk 2]
```diff
@@ -84,12 +84,14 @@ private int testDefault(int testnum) throws Exception {
...
         doTest(c, testnum++);
 
         // get an instance of JavaLoginConfig from SunRsaSign
         try {
-            c = Configuration.getInstance(JAVA_CONFIG, null, "SunRsaSign");
+            c = Configuration.getInstance(JAVA_CONFIG, null,
+                    System.getProperty("test.provider.name","SunRsaSign"));
```
Comment (turbanoff) [inline test/jdk/javax/security/auth/login/Configuration/GetInstance.java:94]:
```suggestion
                    System.getProperty("test.provider.name", "SunRsaSign"));
```

**Comment by seanjmullan** [comment]
You are changing more than SunJCE providers, so the title of this bug should not be specific to SunJCE. Suggest: "Replace hardcoded security providers with new test.provider.name system property".

Are there any cases where a test has more than one hardcoded provider?

**Comment by mpdonova** [comment]
> You are changing more than SunJCE providers, so the title of this bug should not be specific to SunJCE. Suggest: "Replace hardcoded security providers with new test.provider.name system property".
> 
> Are there any cases where a test has more than one hardcoded provider?

I changed the name of the bug and PR. Yes there are cases where a test has more than one hardcoded provider.

**Comment by seanjmullan** [comment]
> I changed the name of the bug and PR. Yes there are cases where a test has more than one hardcoded provider.

How would that work then if the property only allows you to specify one provider?

**Comment by mpdonova** [comment]
> > I changed the name of the bug and PR. Yes there are cases where a test has more than one hardcoded provider.
> 
> How would that work then if the property only allows you to specify one provider?

The tests with multiple, hardcoded providers do that because the SunJCE provider doesn't implement every engine/algorithm. I've seen this mostly with Signatures and KeyPairGenerators. A provider specified via the property would have to implement all of the operations used in the test.

**Review by ascarpino — APPROVED** [review]

File: doc/testing.html [hunk 3]
```diff
@@ -586,6 +592,18 @@ <h3 id="pkcs11-tests">PKCS11 Tests</h3>
     JTREG=&quot;JAVA_OPTIONS=-Djdk.test.lib.artifacts.nsslib-linux_aarch64=/path/to/NSS-libs&quot;</code></pre>
 <p>For more notes about the PKCS11 tests, please refer to
 test/jdk/sun/security/pkcs11/README.</p>
+<h3 id="testing-with-external-security-providers">Testing with external
+security providers</h3>
```
Comment (seanjmullan) [inline doc/testing.html:596]: Avoid the word "external". Change this to "Testing with an alternate security provider".

File: doc/testing.html [hunk 4]
```diff
@@ -586,6 +592,18 @@ <h3 id="pkcs11-tests">PKCS11 Tests</h3>
...
+<p>Some security tests currently use a hardcoded provider for
+KeyFactory, Cipher, KeyPairGenerator, KeyGenerator, or SecretKeyFactory.
+Specify the <code>-Dtest.provider.name=NAME</code> property, to use a
+specific provider when instantiating the service.</p>
```
Comment (seanjmullan) [inline doc/testing.html:600]: s/specific/different/
No comma needed after "property".
s/service/service(s)/

File: doc/testing.html [hunk 5]
```diff
@@ -586,6 +592,18 @@ <h3 id="pkcs11-tests">PKCS11 Tests</h3>
     JTREG=&quot;JAVA_OPTIONS=-Djdk.test.lib.artifacts.nsslib-linux_aarch64=/path/to/NSS-libs&quot;</code></pre>
 <p>For more notes about the PKCS11 tests, please refer to
 test/jdk/sun/security/pkcs11/README.</p>
+<h3 id="testing-with-external-security-providers">Testing with external
+security providers</h3>
+<p>Some security tests currently use a hardcoded provider for
```
Comment (seanjmullan) [inline doc/testing.html:597]: Remove "currently".
s/for/when instantiating/
s/or SecretKeyFactory/or SecretKeyFactory objects/

File: doc/testing.html [hunk 6]
```diff
@@ -586,6 +592,18 @@ <h3 id="pkcs11-tests">PKCS11 Tests</h3>
...
+KeyFactory, Cipher, KeyPairGenerator, KeyGenerator, or SecretKeyFactory.
```
Comment (seanjmullan) [inline doc/testing.html:598]: Put the APIs in code font.

File: doc/testing.md [hunk 7]
```diff
@@ -603,6 +603,20 @@ $ make test TEST="jtreg:sun/security/pkcs11/Secmod/AddTrustedCert.java" \
 For more notes about the PKCS11 tests, please refer to
 test/jdk/sun/security/pkcs11/README.
 
+### Testing with external security providers
```
Comment (seanjmullan) [inline doc/testing.md:606]: Same comments as above apply here.

File: doc/testing.md [hunk 8]
```diff
@@ -603,6 +603,20 @@ $ make test TEST="jtreg:sun/security/pkcs11/Secmod/AddTrustedCert.java" \
...
+
+Some security tests currently use a hardcoded provider for KeyFactory, Cipher,
+KeyPairGenerator, KeyGenerator, or SecretKeyFactory. Specify the
+`-Dtest.provider.name=NAME` property, to use a specific provider when
+instantiating the service.
+
+### Testing with external security providers
```
Comment (seanjmullan) [inline doc/testing.md:613]: Duplicate entry?

**Review by seanjmullan — CHANGES_REQUESTED** [review]

**Review by rhalade — APPROVED** [review]

File: doc/testing.html [hunk 9]
```diff
@@ -592,18 +589,14 @@ <h3 id="pkcs11-tests">PKCS11 Tests</h3>
...
-<p>Some security tests currently use a hardcoded provider for
-KeyFactory, Cipher, KeyPairGenerator, KeyGenerator, or SecretKeyFactory.
-Specify the <code>-Dtest.provider.name=NAME</code> property, to use a
-specific provider when instantiating the service.</p>
+<h3 id="testing-with-alternative-security-providers">Testing with
+alternative security providers</h3>
+<p>Some security tests use a hardcoded provider for
+<code>KeyFactory</code>, <code>Cipher</code>,
+<code>KeyPairGenerator</code>, <code>KeyGenerator</code>, or
+<code>SecretKeyFactory</code> objects. Specify the
```
Comment (seanjmullan) [inline doc/testing.html:597]: I see some tests use a hard-coded provider for `Signature` too. Can you double-check all the changes and add any that you missed?

**Review by seanjmullan — APPROVED** [review]
Approving but I recommend a follow-on task be filed to check if some of these hard-coded providers are really necessary.

**Comment by RealLucy** [comment]
While reviewing the backport PR to jdk17, I stumbled across a potential functional incompleteness. There are a few tests which use more than one security provider. This works well when only the default providers are used. One example of such a test is `test/jdk/java/security/Signature/TestCloneable.java`. There are three providers by default:
```
        String dsaProviderName = System.getProperty("test.provider.name", "SUN");
        String ecProviderName  = System.getProperty("test.provider.name", "SunEC");
        String rsaProviderName = System.getProperty("test.provider.name", "SunRsaSign");
```
If the `test.provider.name` property is specified for a test run, all three name strings are set to the same provider name. The test will only succeed if the specified provider can handle the combined set of algorithms of all three default providers. 

These are the tests I found to be affected. The list is **not** believed to be complete.
```
test/jdk/java/security/Signature/TestCloneable.java
test/jdk/java/security/SignedObject/Chain.java
test/jdk/java/security/KexGenerator/TestGetInstance.java
test/jdk/java/security/Mac/TestGetInstance.java
test/jdk/java/security/auth/login/Configuration/GetInstance.java
test/jdk/java/security/X509/X509CertImpl/Verify.java
```

**Pull request closed without merging**